### ISSUE TITLE
way to prevent infinite loops?

### ISSUE THREAD

**Issue by mspanish** [Taskkill/snippet_runner_js#1]
Hello your module works quite well - I've combined it with a Prism live editing plugin and the results are very good. One problem is that the user can write an infinite loop which causes the browers to crash. Is there a way to prevent that?

thanks very much!

**Comment by Taskkill** [comment]
Hello and thank you for using my project :)
I am really sorry, I am so late with my response, I was just finishing exams, hope you don't mind.

In general it sound like non trivial problem, same problem for example occurs with jsfiddle.net or these kind of sandboxes. I will definitely try to find some solution, I will update this issue when something reasonably good occurs to me.
Thank you again for you patience and for taking time to work with my project.

**Comment by mspanish** [comment]
hello and thanks so much!! I will take a look - I already have your script live on my "work in progress"guide to es6 - you can see the result here [](https://es6.kwippe.com/topics/arrow_functions) - I'll be sure to check your most recent version and let you know!

**Comment by Taskkill** [comment]
Nice. That's excellent :) I think new version of this project should work very much out of the box, it has very good documentation in readme too.

As for infinite loops - I think there is one way. Isolating user snippets inside web workers. It wont freeze browser and as long as your snippet will not need global document object, it should work same way.

Your project is exactly the kind of project which was motivation for me to develop this module. I am very pleased it is doing it's job after all.

Don't hesitate to hit me up with anything about newer version, I hope everything will work same way as in original or better.

**Comment by Taskkill** [comment]
I now realize, you maybe don't need to use dynamic-sandbox. I was looking through the sources of page you showed me here and I think, maybe you would prefer just this repository to have this simple enhancement.
I am really able to update this very repository with new functionality.
I don't really want to force you to using dynamic-sandbox, when you don't really want sandboxing and restricting user access to global variables and so. I figured out all you want is something what will execute user code and to mock console. Correct me if I am wrong, otherwise I am OK with updating this project or the other one, it's up to you.

**Comment by mspanish** [comment]
hey yes you are right - I do need access to things like Object and Array - I tried JSON.stringify and couldn't get that working either, so i did an override in my own code that if the output is [object Object] I stringify it!

sorry I didn't reply yet been busy with kids stuff! But I really thank you for doing this, the other solutions are hugely bloated like CodeIgnighter, Ace, and even Runkit is large and not easy to implement in this fashion.  With all of the need for interactive lessons yours addresses a big need right now!

**Comment by Taskkill** [comment]
Yeah, I saw that in your code.
Well then, I will update this repo to fit you needs. I will also write sufficient documentation to readme, this way you will be able to get JSON.stringify to work and use it in simpler fashion. It should be done soon.
Thanks again for your appreciation and patience.

**Comment by mspanish** [comment]
Hey you are awesome!  I hope to get back to my es6 project in a couple  of days - I'll be including a tech stack page where I thank people and reference all the libraries used - and will be sure to credit you there.  Hope to see more repos from you, you're obviously an awesome JS'er :)

**Comment by Taskkill** [comment]
Thank you for your kind words.
I have been doing some research in my free time and I think I found out, how exactly solve the problem.
Unfortunately I cannot implement it, because it depends on `SharedWorker` `SharedArrayBuffer` and `Atomics`.

`details ahead, you can skip them`

Main idea: I will create SharedWorker which will receive your snippet in message and then executes it, whenever your snippet requests variable, for example some function, which is trapped or being intercepted, such as your custom `console.log` or `console.err`, my code will instead pass some custom function, which will act as middleware, it will receive arguments for trapped function, then it will send request for calling actual function you supplied to my API, but this actual function from you, must and will be invoked in main thread, then it will use `Atomics.wait`on mutex to interrupt worker thread and wait till main thread executes your function (for example your fake `console.log` or any other fn), then it will post message to `SharedWorker` with appropriate resulting value in case your function returns. This value will then be stored inside worker thread and main thread will wake worker thread using `Atomics.notify` on mutex. Worker thread will then wake and get the returning value, this value will be returned from my middleware, where execution was stopped til now, back to your snippet. Then your snippet will continue in execution, again able to invoke functions from absolutely separated thread without losing scope and closures (you really need this, for your fake console for example) and same with other variables.

TLDR: I need to use advanced JS and browser features to implement predictable and completely valid parallelism with interruptions and notifications - basically mutexes. I don't know about better and simpler approach.

Unfortunately one of the main features `SharedArrayBuffer` is disabled due to Spectre vulnerabilities. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#Syntax

So, until it will be re-enabled I can't really do anything. Sorry.
Once it will be re-enabled I will implement it promptly.

In the meantime I simplified my code and updated API, you can see in README, where I showed very basic usage, if you use it this 'new' way, you can use `console.log(JSON.stringify(stuff) )` in your snippets. Let me know in case you need some more examples or adjustments.
Thank you for your patience. I hope it will help.

**Comment by Taskkill** [comment]
Yeah, I am afraid I am little bit clumsy at explaining this stuff :smile: 
I saw it and really liked it, hope many people will use it to learn more interactively and the fun way.
